src/workers/personalizedDigestEmail.ts

```diff
@@ -0,0 +1,178 @@
+    return {
+      post_title: post.title,
+      post_image: post.image || pickImageUrl(post),
+      post_link: addNotificationUtm(
+        getDiscussionLink(post.id),
+        'email',
+        'digest',
+      ),
+      source_name: post.sourceName,
+      source_image: post.sourceImage,
+    };
+  });
+
+  return templateData;
+};
+
+const worker: Worker = {
+  subscription: 'api.personalized-digest-email',
+  handler: async (message, con, logger) => {
+    if (process.env.NODE_ENV === 'development') {
+      return;
+    }
+
+    const data = messageToJson<Data>(message);
+    const { personalizedDigest } = data;
+
+    const user = await con.getRepository(User).findOneBy({
+      id: personalizedDigest.userId,
+    });
+
+    if (!user?.infoConfirmed) {
+      return;
+    }
+
+    const currentDate = new Date();
+    const emailSendDate = getEmailSendDate({ personalizedDigest });
+    const previousSendDate = getPreviousSendDate({ personalizedDigest });
+
+    const feedConfig = await feedToFilters(
+      con,
+      personalizedDigest.userId,
+      personalizedDigest.userId,
+    );
+
+    const feedResponse = await personalizedDigestFeedClient.fetchFeed(
+      { log: logger },
+      personalizedDigest.userId,
+      {
+        user_id: personalizedDigest.userId,
+        total_posts: personalizedDigestPostsCount,
+        date_from: format(previousSendDate, personalizedDigestDateFormat),
+        date_to: format(currentDate, personalizedDigestDateFormat),
+        allowed_tags: feedConfig.includeTags,
+        blocked_tags: feedConfig.blockedTags,
+        blocked_sources: feedConfig.excludeSources,
+      },
+    );
+
+    const posts: TemplatePostData[] = await fixedIdsFeedBuilder(
+      {},
+      feedResponse.map(([postId]) => postId),
+      con
+        .createQueryBuilder(Post, 'p')
+        .select(
+          'p.id, p.title, p.image, p."createdAt", s.name as "sourceName", s.image as "sourceImage"',
+        )
+        .leftJoin(Source, 's', 'p."sourceId" = s.id'),
+      'p',
+    ).execute();
+
+    if (posts.length === 0) {
+      logger.warn(
+        { data: messageToJson(message) },
+        'no posts found for personalized digest',
+      );
+
+      return;
+    }
+
+    const [dayName] = Object.entries(DayOfWeek).find(
+      ([, value]) => value === personalizedDigest.preferredDay,
+    );
+    const userName = user.name?.trim().split(' ')[0] || user.username;
+    const emailPayload: MailDataRequired = {
+      ...baseNotificationEmailData,
+      to: {
+        email: user.email,
+        name: user.name,
+      },
+      sendAt: Math.floor(emailSendDate.getTime() / 1000),
+      templateId: emailTemplateId,
+      dynamicTemplateData: {
+        day_name: dayName,
+        first_name: userName,
+        posts: await getPostsTemplateData({ posts }),
+      },
+      asm: {
+        groupId: 23809,
+      },
+      category: 'Digests',
+    };
+
+    await sendEmail(emailPayload);
```

Comment:
@idoshamun should we wrap this in try catch, I am worried if sendgrid fails this would retry, in development it retries forever, not sure how many retries it does in production?

Reply:
I don't think we have a limit on number of retries currently, would you like to check?

Reply:
If it's not possible, I suggest to try and catch. It's better not to ddos sendgrid much more than we will 😂

Reply:
Haha ok, for sure we don't want infinite loops/retries again this week 😆 I'll take a look

Reply:
@idoshamun I think for solving retries we need to setup dead lettering https://cloud.google.com/pubsub/docs/handling-failures#dead_letter_topic
In that topic we can for now just log the message payload and exit. Dead letter topic can be enabled per subscription from what I can see so it does not need to be system wide change. Could we setup something like this?

As for solving stuff for this PR I added try catch around the whole resolver so we make sure if anything goes wrong we don't have thousands of messages retrying forever.

Reply:
You can find here an example of a dead letter provisioning with pulumi:
https://github.com/dailydotdev/analytics-processor/blob/main/.infra/index.ts#L45

Reply:
Merging this now and will look into dead letter topic after.